File: app/src/main/java/com/ivantha/playtolearn/listener/BoardTouchListener.java
package com.ivantha.playtolearn.listener;

import android.content.Context;
import android.view.MotionEvent;
import android.view.View;

import com.ivantha.playtolearn.activity.BoardActivity;
import com.ivantha.playtolearn.widget.ProblemDialog;

import java.io.Serializable;

public class BoardTouchListener implements View.OnTouchListener, Serializable {
    private Context context;

    private float mDownX;
    private float mDownY;
    private final float SCROLL_THRESHOLD = 10;
    private boolean isOnClick;

    public BoardTouchListener(Context context) {
        this.context = context;
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mDownX = event.getX();
                mDownY = event.getY();
                isOnClick = true;
                break;
            case MotionEvent.ACTION_UP:
                if (isOnClick) {
                    BoardActivity.BoardTileLayout boardTileLayout;
                    if (v instanceof BoardActivity.BoardTileLayout) {
                        boardTileLayout = (BoardActivity.BoardTileLayout) v;
                        if (boardTileLayout.getBoardTile().getQuestion() != null) {
                            showQuestionDialog(context, boardTileLayout.getBoardTile().getQuestion().getTitle(),
                                    boardTileLayout.getBoardTile().getQuestion().getDescription());
                        }
                    } else if (v instanceof BoardActivity.BoardTileButton) {
                        boardTileLayout = (BoardActivity.BoardTileLayout) v.getParent();
                        showQuestionDialog(context, boardTileLayout.getBoardTile().getQuestion().getTitle(),
                                boardTileLayout.getBoardTile().getQuestion().getDescription());
                    }
                }
                break;
            case MotionEvent.ACTION_MOVE:
                if (v instanceof BoardActivity.BoardTileButton) {
                    if (isOnClick && (Math.abs(mDownX - event.getX()) > SCROLL_THRESHOLD || Math.abs(mDownY - event.getY()) > SCROLL_THRESHOLD)) {
                        View.DragShadowBuilder shadowBuilder = new View.DragShadowBuilder(v);
                        v.startDrag(null, shadowBuilder, v, 0);
                        isOnClick = false;
                    }
                }
                break;
            default:
                break;
        }

        return true;
    }

    public static void showQuestionDialog(Context context, String title, String description){
        ProblemDialog problemDialog = new ProblemDialog(context);
        problemDialog.setTitle(title);
        problemDialog.setDescription(description);
        problemDialog.show();
    }
}

File: app/src/main/java/com/ivantha/playtolearn/listener/BoardTouchListener.kt
package com.ivantha.playtolearn.listener

import android.content.Context
import android.view.MotionEvent
import android.view.View

import com.ivantha.playtolearn.widget.BoardTileButton
import com.ivantha.playtolearn.widget.BoardTileLayout
import com.ivantha.playtolearn.widget.ProblemDialog

import java.io.Serializable

class BoardTouchListener(private val context: Context) : View.OnTouchListener, Serializable {

    private var mDownX: Float = 0.toFloat()
    private var mDownY: Float = 0.toFloat()
    private val scrollThreshold = 10f
    private var isOnClick: Boolean = false

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                mDownX = event.x
                mDownY = event.y
                isOnClick = true
            }
            MotionEvent.ACTION_UP -> if (isOnClick) {
                val boardTileLayout: BoardTileLayout
                if (v is BoardTileLayout) {
                    boardTileLayout = v
                    if (boardTileLayout.tile.question != null) {
                        showQuestionDialog(context, boardTileLayout.tile.question!!.title,
                                boardTileLayout.tile.question!!.description)
                    }
                } else if (v is BoardTileButton) {
                    boardTileLayout = v.getParent() as BoardTileLayout
                    showQuestionDialog(context, boardTileLayout.tile.question!!.title,
                            boardTileLayout.tile.question!!.description)
                }
            }
            MotionEvent.ACTION_MOVE -> if (v is BoardTileButton) {
                if (isOnClick && (Math.abs(mDownX - event.x) > scrollThreshold || Math.abs(mDownY - event.y) > scrollThreshold)) {
                    val shadowBuilder = View.DragShadowBuilder(v)
                    v.startDrag(null, shadowBuilder, v, 0)
                    isOnClick = false
                }
            }
        }

        return true
    }

    companion object {
        fun showQuestionDialog(context: Context, title: String?, description: String?) {
            val problemDialog = ProblemDialog(context)
            problemDialog.setTitle(title!!)
            problemDialog.setDescription(description!!)
            problemDialog.show()
        }
    }
}
